### ISSUE TITLE
Parsing URLs in freeform text

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [ELLIOTTCABLE/excmd.js#4]
There's some decisions to make regarding the parsing of bare URLs in freeform text.

Worth noting is that whatever decisions we make here, the user can still exceed these limitations by entering strange URLs as a string — for instance, URLs containing spaces, or vertical-bars (which are pretty much precluded by support for command-composition.)

Also worth noting why we shouldn't support *any* URL in freeform text: at that point, basically *any* character valid in a URL, becomes a character we can't assign other meaning to reliably. Think `|` or `/` or `:` being available for further semantics, allowing the language more expressiveness overall.

That said, let's chat about how I'm going to support URLs entered *without* quotes.

### How ‘a URL’ is introduced by the user

First off, I've started adding a separate `URL` mode to our [modal lexer](https://github.com/ELLIOTTCABLE/excmd.js/blob/master/src/lexer.body.ml#L263-L269). Once this mode is introduced (see below), a much larger set of characters are accepted as a continued part of that identifier, before breaking back out into normal running text. (Basically, ‘recognized URLs’ are low-key strings.)

First, introductory sequences. How do we decide when we're probably seeing a URL?

1. ```
   http://google.com
   ```

   The simplest case, and easy to parse in running text, is when a collection of quantifiers exist: a known protocol followed by a colon (`http:`), the magic colon-slash-slash sequence (`://`), or an extremely popular TLD (`.com`). These are obviously a minimal starting point, the least required of the lexer in running text.

2. ```
   magnet:?xt=urn:btih:c12fe1c06bba254a9dc9f519b335aa7c1367a88a
   file:/a/path
   weird-widget://google.com/search
   google.com/search
   ```

   In some cases, we have *some* reasonably-deterministic heuristics that we can safely say “hey, if you make this look like a URL, <future feature depending on `/`> won't work, because we wanted to support more URLs way-back-when. Put spaces around it.”

   Personally, here's the list of things I think I should be supporting (i.e. the initial components that introduce a URL, and allow for the additional URL components listed further below)

   1. Any (`[a-z]+`) sequence followed by a colon. (*Because* URLs are so important to us, and this is such an excellent way to recognize such a large set of possible URLs, I've opted to disallow this character elsewhere in the lexical analysis.) This catches slightly-less-common URL protocol components like `magnet` or `urn`, and is compatible with the ‘initial colon’ of a Vim excmd, which can only appear at the start of a command, after a newline, or after a separator like `|` or `;` … never after an ASCII letter. `magnet:?xt=urn:btih:c12fe1c0...

      We *could* also limit this to a determined subset of known protocols. It's not exactly like there's a huge array of protocols in everyday use that *don't* dictate `://`, which below would handle. `magnet`, `urn`, `file` come to mind …

      This is one of the primary reasons I open this Issue to discussion, btw. I'm a little torn here — restricting this would allow considerable flexibility in the future.

   2. Anything that's expressed elsewhere in the grammar as a valid [UAX#31 internationalized identifiers](http://unicode.org/reports/tr31/#A1), followed by a `://`. This precludes ever allowing that sequence of characters anywhere else, and I think that's a Good Decision; this should handle a ton of internationalized or weird proprietary-tech use-cases, and doesn't really lose us any flexibility in implementation. `weird-widget://google.com/foo`

   3. Valid UAX#31 identifiers *ending* in a subset of well-known TLDs, `.com`, `.net`, `.co.uk` … this is obviously a ballooning target to meet, but I definitely think it's worth our time, because it allows us to catch things that the user is going to think are "obvious", like typing in `google.com/search...`.

   Finally, it's worth noting that anything that parses as an IDENTIFIER is implicitly valid here as well — don't forget that the purpose of this is to catch things that are *not* valid IDENTIFIERs, like `google.com/search` (due to the slash), but that we want the user to be able to enter bare. For instance, `pixijs.download` and `日本.icom.museum` are supported implicitly by the lexer, without any special handling.

3. ```
   weird-widget:foo
   pixijs.download/release/docs
   ```

   In terms of things *explicitly* outside our supported set, I thing the following belong — both for consistency, and for future expandability:

   1. Protocols that A. are not in the set of ones we hard-code, B. don't use `://`, C. *and* don't use a well-known domain-name. (Supporting those basically means using a colon triggers URL-special-casing in every possible circumstance, and precludes, among other things, supporting ternary operators at a later date. I don't think that edge-case is worth the tradeoff; a user can just put these weird outliers in quotation.) eg. `"bleh:fleb#woggle"` would have to be quoted.

   2. Path-holding URLs without *any* of the above qualifiers. See below, but basically, I want to reserve various URL-compatible special characters for future extensions to the language; and again, this is an uncommon edge-case that I think we're safe excluding this from the lexical analyses. eg. `"iloveweirdTLDs.boats/resume.html"` would need to be in quotes.

I think the above covers a very large subset of the most common URLs that users might be entering by hand, while maintaining reasonable restrictions on the lexical language — feedback is welcome, though!

### What's valid inside ‘a URL’, once introduced

Now, once ‘URL mode’ is triggered by one of the above, any feature invoked by a character that conflict with URL-compatible characters would have to be triggered by whitespace ... *almost*. (Specifically, I want to allow vertical-bar and unbalanced delimiters to break out of URL-mode without whitspace — but I'll come back to that.)

For instance, in the middle of a URL, I plan to accept the following outside quotation (except with some Unicode characters supplanting the usual `A-Za-z` ... I think? probably the same ones we allow in identifiers ...):


```javascript
/[A-Za-z0-9\-._~:/?#[\]@!$&()*+,;=]/
```

However, I want to add two restrictions:

1. Some of those characters may be restricted in "final position", i.e. they may be rejected as the last character before spacing / EOF / whatever. I'm planning on accepting:

   ```javascript
   /[A-Za-z0-9\-._/\])+]/ // subset of the above *accepted* in final position
   /[~:?#[@!$&(*,;=]/     // subset of the above *rejected* in final position
   ```

   (Adjustments to these are very welcome, I came up with them basically arbitrarily ...)

2. Delimiters in freeform-text URLs *must be balanced*. That is, you can include a closing delimiter **iff** you've included a opening delimiter below it. That ensures that future constructs like the following don't require additional disambiguating whitespace:

       foo --bar (bindurl google.com/search)
       bindurl en.wikipedia.org/wiki/Java_(programming_language) // but this still works, too

----

Okay! Phew! Thanks for reading (or more likely, skimming!) all this — to be honest, it's a lot more straight-forward in code; but I wanted to include rationatle and something that could form the seeds of eventual end-user documentation here.

**tl;dr**: User should be able to type almost anything that looks like a URL into freeform text without quotes. This is subordinate to retaining our flexibility with future features, but only in a couple edge-cases.

**Comment by bovine3dom** [comment]
Related Tridactyl issue to make this easier to find https://github.com/tridactyl/tridactyl/issues/789

I haven't had time to read through this in detail, and probably won't for a while, but I think cmcaine was on the right track when he said we should just test it on a corpus and see how well it does. A very cursory search led me to, e.g, https://github.com/h2o/h2o/tree/master/fuzz/url-corpus

**Comment by antonva** [comment]
Nice writeup!
I read this a couple of times today and the first 2 points I have are:

* I believe a lot of URL encoded strings would get sad if the `%` character was not allowed "inside a URL".
* (3.ii) It'd be wrong of us to require people to use quotes around URLs we decide aren't popular enough to be real URLs.

**Comment by ELLIOTTCABLE** [comment]
> Nice writeup!
> I read this a couple of times today and the first 2 points I have are:
> 
>     * I believe a lot of URL encoded strings would get sad if the `%` character was not allowed "inside a URL".

Oh! I didn't even include that section — percent-encoding is handled separately, basically. Or was at the time that I wrote the above. Either way, yes, support for that is definitely a given! Oops! 🤣

>     * (3.ii) It'd be wrong of us to require people to use quotes around URLs we decide aren't popular enough to be real URLs.

You talking about TLDs? Easy enough to try and raise that coverage arbitrarily high (‘top’ 100 TLDs? top 200?).

Another alternative is to requisition the `.` entirely for URL-mode — i.e. any ‘identifier’ with a `.` in it, must then be separated from eventual special operators (`-+/`, maybe more) by whitspeace or quotation. That discards domain-names from this discussion entirely, and handles the vast majority of URLs easily! (Actually, I quite like this option, now that I think on it.)

**Comment by cmcaine** [comment]
Dots is what I was thinking too. We can still use them in arguments, etc,
they're just not part of the grammar.

**Comment by cmcaine** [comment]
I mean, I was thinking it about forward slash, but dots are probably better.

I don't think it matters much if the URL detection code is a bit greedy if
you can easily disable it with quotes or whitespace.

**Comment by ELLIOTTCABLE** [comment]
> I mean, I was thinking it about forward slash, but dots are probably better. I don't think it matters much if the URL detection code is a bit greedy if you can easily disable it with quotes or whitespace.

Whitespace, yeah — note though that quotes are the opposite (i.e. one could disambiguate something *away* from a given special meaning, into an IDENTIFIER/URL-equivalent that's valid in WORD-position in a command, using quotes. The only way to escape URL-mode, though, is whitespace!)

Thanks for the feedback, all. Time to switch the implementation over to dots; lets see if that fixes my exploding-complexity issue, too!

Oh — and a corpus of URLs to test the lexer against is a great idea; maybe somebody could snag gluing that together as a task for y'all's meetup this spring. =)

**Comment by cmcaine** [comment]
The other thing we should be sure to handle well is escape to js.

JS has loads of dots and other special chars and we should have some good
syntax for escaping to js mode where that makes sense.

So backtick quotes or whatever should beat the URL parsing.
